agent/app/agent_capability.go

```diff
@@ -316,6 +345,92 @@ func (agent *ecsAgent) appendTaskENICapabilities(capabilities []*ecs.Attribute)
 	return capabilities
 }
 
+func (agent *ecsAgent) appendExecCapabilities(capabilities []*ecs.Attribute) ([]*ecs.Attribute, error) {
+	// for an instance to be exec-enabled, it needs resources needed by SSM (binaries, configuration files and certs)
+	// the following bind mounts are defined in ecs-init and added to the ecs-agent container
+
+	capabilityExecRootDir := filepath.Join(capabilityDepsRootDir, capabilityExec)
+	binDir := filepath.Join(capabilityExecRootDir, capabilityExecBinRelativePath)
+	configDir := filepath.Join(capabilityExecRootDir, capabilityExecConfigRelativePath)
+	certsDir := filepath.Join(capabilityExecRootDir, capabilityExecCertsRelativePath)
+
+	// top-level folders, /bin, /config, /certs
+	dependencies := map[string][]string{
+		binDir:    []string{},
+		configDir: []string{},
+		certsDir:  capabilityExecRequiredCerts,
+	}
+	if exists, err := dependenciesExist(dependencies); err != nil || !exists {
+		return capabilities, err
+	}
+
+	// ssm binaries are stored in /bin/<version>/, 1 version is downloaded by ami builder for ECS instances
+	binDependencies := map[string][]string{}
+	// child folders named by version inside binDir, e.g. 3.0.236.0
+	binFolders, err := getSubDirectories(binDir)
+	if err != nil {
+		return capabilities, err
+	}
+	for _, binFolder := range binFolders {
+		// check for the same set of binaries for all versions for now
+		// change this if future versions of ssm agent require different binaries
+		versionSubDirectory := filepath.Join(binDir, binFolder)
+		binDependencies[versionSubDirectory] = capabilityExecRequiredBinaries
+	}
+	if len(binDependencies) < 1 {
+		return capabilities, nil
+	}
+	if exists, err := dependenciesExist(binDependencies); err != nil || !exists {
```

Comment:
I'd argue that we need to find at least one valid dependencies dir, as opposed to all of them. In the future when we create an update process, this process might be in progress and not have all the dependencies right at the time of our check, which would cause this to error out.

Reply:
In other words, only error out if no valid versions were found.

Reply:
good point, change to error out only if no valid version is found

Reply:
good point, add deny-list for ssm versions; currently this will be an empty set so no actual effect before we add any version into it